sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -1636,6 +1638,30 @@ object ReplaceExceptWithAntiJoin extends Rule[LogicalPlan] {
 }
 
 /**
+ * Insert a [[Scanner]] operator over [[MultiInstanceRelation]], then combine [[Project]]s and
+ * [[Filter]]s with this operator, then the planner can match on the [[Scanner]] node directly.
+ */
+object InsertRelationScanner extends Rule[LogicalPlan] with PredicateHelper {
```

Comment:
This doesn't reuse the existing `ColumnPruning` and `PushDownPredicate` rules. I'd like to add the wrapper at the very beginning, e.g. `SessionCatalog.lookUpRelation`.